snapshotter/snapshotter/pull.go

```diff
@@ -0,0 +1,267 @@
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// Fetch image from reference
+func fetchImage(image string) (img v1.Image, err error) {
+	ref, err := name.ParseReference(image)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse image reference: %s. %w", image, err)
+	}
+
+	img, err = remote.Image(ref)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch image %q, make sure it exists. %w", image, err)
+	}
+	conf, _ := img.ConfigName()
+	log.Debugf("Image id: %s", conf.String())
+	return
+}
+
+func layerDiskImageName(chainId string) string {
+	return strings.Split(chainId, ":")[1] + ".disk"
+}
+
+func layerManifestName(chainId string) string {
+	return strings.Split(chainId, ":")[1] + ".json"
+}
+
+func moveFile(src string, dest string) error {
+	// Try renaming file
+	err := os.Rename(src, dest)
+	if err == nil {
+		return nil
+	}
+
+	// Rename failed. Copy file
+	in, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+
+	out, err := os.Create(dest)
+	if err != nil {
+		return err
+	}
+
+	_, err = io.Copy(out, in)
+	return err
+}
+
+// Download a given layer and convert it to a disk image. Return the root hash.
+func downloadLayer(layerChainId string, layerNumber int, layer v1.Layer, storagePath string) (string, error) {
+	diskImageName := layerDiskImageName(layerChainId)
+	diskImagePath := filepath.Join(storagePath, diskImageName)
+
+	// Check if the layer has already been downloaded.
+	if _, err := os.Stat(diskImagePath); err == nil {
+		rootHash, err := ReadRootHash(diskImagePath)
+		if err == nil {
+			fmt.Printf("Layer #%d, skipping pull of %s\n", layerNumber, diskImagePath)
+			return rootHash, nil
+		}
+	}
+
+	rc, err := layer.Uncompressed()
+	if err != nil {
+		return "", fmt.Errorf("failed to uncompress layer %s. %w", layerChainId, err)
+	}
+
+	// Use a temporary file for the download.
+	tmpfile, err := ioutil.TempFile("", diskImageName)
+	if err != nil {
+		return "", fmt.Errorf("failed to temporary file for %s. %w", layerChainId, err)
+	}
+	tmpName := tmpfile.Name()
+	defer os.Remove(tmpfile.Name())
+
+	log.Debug("converting tar to layer disk image")
+	rootHash, err := TarStreamToDiskImage(rc, tmpName)
+	if err != nil {
+		return "", fmt.Errorf("failed to convert tar to ext4. %w", err)
+	}
+
+	if err := moveFile(tmpName, diskImagePath); err != nil {
+		// Check if the file has been created.
+		if _, err1 := os.Stat(diskImagePath); err1 != nil {
+			return "", fmt.Errorf("failed to rename %s to %s. %w", tmpName, diskImagePath, err)
+		}
+	}
+
+	fmt.Fprintf(os.Stdout, "Pulled Layer %d: %s\nRootHash=%s\n", layerNumber, diskImagePath, rootHash)
+	return rootHash, nil
+}
+
+func digest(str string) string {
+	s := sha256.Sum256([]byte(str))
```

Comment:
Ah, so here is the use of SHA-256. I got confused earlier by MD5 usage in the superblock.